errors.go

```diff
@@ -0,0 +1,95 @@
+package flux
+
+import (
+	"encoding/json"
+	"errors"
+)
+
+type HelpfulError interface {
+	Base() *BaseError
+}
+
+// Representation of errors in the API. These are divided into a small
+// number of categories, essentially distinguished by whose fault the
+// error is; i.e., is this error:
+//  - a transient problem with the service, so worth trying again?
+//  - not going to work until the user takes some other action, e.g., updating config?
+
+type BaseError struct {
+	// a message that can be printed out for the user
+	Help string `json:"help"`
+	// the underlying error that can be e.g., logged for developers to look at
+	Err error
+}
+
+func (e *BaseError) Base() *BaseError {
+	return e
+}
+
+func (e *BaseError) Error() string {
+	return e.Err.Error()
+}
+
+func (e *BaseError) MarshalJSON() ([]byte, error) {
+	var errMsg string
+	if e.Err != nil {
+		errMsg = e.Err.Error()
+	}
+	jsonable := &struct {
+		Help string `json:"help"`
+		Err  string `json:"error,omitempty"`
+	}{
+		Help: e.Help,
+		Err:  errMsg,
+	}
+	return json.Marshal(jsonable)
+}
+
+func (e *BaseError) UnmarshalJSON(data []byte) error {
+	jsonable := &struct {
+		Help string `json:"help"`
+		Err  string `json:"error,omitempty"`
+	}{}
+	if err := json.Unmarshal(data, &jsonable); err != nil {
+		return err
+	}
+	if jsonable != nil {
+		e.Help = jsonable.Help
+		if jsonable.Err != "" {
+			e.Err = errors.New(jsonable.Err)
+		}
+	}
+	return nil
+}
+
+func CoverAllError(err error) *BaseError {
+	return &BaseError{
+		Err: err,
+		Help: `An error occured for which we don't have a specific message.
+
+If you see this, it means we need to come up with a better message! It
+would help us if you log an issue at
+https://github.com/weaveworks/flux/issues saying what you were doing
+when you saw this, and quoting the following:
+
+    ` + err.Error(),
```

Comment:
I'd favour this more if the actual error message were more prominent, and how to report it was more secondary... maybe? Need an example of where this would come up...